File: bumperbot_py_examples/bumperbot_py_examples/simple_parameter.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter

class simpleParameter(Node):
    def __init__(self):
        super().__init__("anand_parameter")

        self.declare_parameter("simple_int_param",28) #name of param with the default value
        self.declare_parameter("simple_string_param","Anand") 

        self.add_on_set_parameters_callback(self.param_change_callback) #when parameter value is changed
        # parameter value can be changed while code is running!

    def param_change_callback(self,params):
        # receives new value of parameter
        result = SetParametersResult() #standard ros2 interface 

        for param in params: # loop through all parameters
            # verfify name and type of parameter!
            if param.name == "simple_int_param" and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info("Param simple_int_param changed! new value is %d: "% param.value)
                result.successful = True #change has been correctly applied to param
            
            if param.name == "simple_string_param" and param.type_ == Parameter.Type.STRING:
                self.get_logger().info("Param simple_string_param changed! new value is %s: "% param.value)
                result.successful = True #change has been correctly applied to param
            
        return result
    

def main():
    rclpy.init()
    simple_parameter = simpleParameter()
    rclpy.spin(simple_parameter)
    simple_parameter.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: bumperbot_cpp_examples/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>

#include <string>
#include <vector>
#include <memory>

using std::placeholders::_1;

class simpleParamter: public rclcpp::Node{
public:
    simpleParamter() : Node("anand_parameter")
    {
        declare_parameter<int>("simple_int_param",28);
        declare_parameter<std::string>("simple_string_param","Anand");

      param_callback_handle_ =  add_on_set_parameters_callback(std::bind(&simpleParamter::param_change_callback,this,_1)); //when param changes
    }    
private:
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
    
    rcl_interfaces::msg::SetParametersResult  param_change_callback(const std::vector<rclcpp::Parameter> & parameters){
            rcl_interfaces::msg::SetParametersResult result;
            for (const auto &param : parameters){
                if (param.get_name()=="simple_int_param" && param.get_type()== rclcpp::ParameterType::PARAMETER_INTEGER){
                    RCLCPP_INFO_STREAM(get_logger(), "Param simple_int_param changed! "<<param.as_int());
                    result.successful = true;

                }
                if (param.get_name()=="simple_string_param" && param.get_type()== rclcpp::ParameterType::PARAMETER_STRING){
                    RCLCPP_INFO_STREAM(get_logger(), "Param simple_string_param changed! "<<param.as_string());
                    result.successful = true;

                }
            }
            return result;
    }
};

int main(int argc, char* argv []){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<simpleParamter>();
    rclcpp::spin(node);
    rclcpp::shutdown();

}